ext/storage_sources/s3_store/s3_store.cpp

```diff
@@ -0,0 +1,247 @@
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <wiredtiger.h>
+#include <wiredtiger_ext.h>
+
+#include <aws/core/Aws.h>
+#include <aws/s3-crt/S3CrtClient.h>
+#include "aws_bucket_conn.h"
+
+#include <iostream>
+
+#define UNUSED(x) (void)(x)
+
+/* S3 storage source structure. */
+typedef struct {
+    WT_STORAGE_SOURCE storage_source; /* Must come first */
+    WT_EXTENSION_API *wt_api;         /* Extension API */
+} S3_STORAGE;
+
+typedef struct {
+    /* Must come first - this is the interface for the file system we are implementing. */
+    WT_FILE_SYSTEM file_system;
+    S3_STORAGE *s3_storage;
+    aws_bucket_conn *conn;
+} S3_FILE_SYSTEM;
+
+/* Configuration variables for connecting to S3CrtClient. */
+const Aws::String region = Aws::Region::AP_SOUTHEAST_2;
+const double throughput_target_gbps = 5;
+const uint64_t part_size = 8 * 1024 * 1024; /* 8 MB. */
+
+/* Setting SDK options. */
+Aws::SDKOptions options;
+
+static int s3_customize_file_system(
+  WT_STORAGE_SOURCE *, WT_SESSION *, const char *, const char *, const char *, WT_FILE_SYSTEM **);
+static int s3_add_reference(WT_STORAGE_SOURCE *);
+static int s3_fs_terminate(WT_FILE_SYSTEM *, WT_SESSION *);
+
+/*
+ * s3_customize_file_system --
+ *      Return a customized file system to access the s3 storage source objects.
+ */
+static int
+s3_customize_file_system(WT_STORAGE_SOURCE *storage_source, WT_SESSION *session,
+  const char *bucket_name, const char *auth_token, const char *config,
+  WT_FILE_SYSTEM **file_systemp)
+{
+    S3_FILE_SYSTEM *fs;
+    int ret;
+
+    /* Mark parameters as unused for now, until implemented. */
+    UNUSED(session);
+    UNUSED(bucket_name);
+    UNUSED(auth_token);
+    UNUSED(config);
+
+    Aws::S3Crt::ClientConfiguration aws_config;
+    aws_config.region = region;
+    aws_config.throughputTargetGbps = throughput_target_gbps;
+    aws_config.partSize = part_size;
+
+    if ((fs = (S3_FILE_SYSTEM*)calloc(1, sizeof(S3_FILE_SYSTEM))) == NULL)
+        return (errno);
+
+    fs->s3_storage = (S3_STORAGE*)storage_source;
+
+    /* New can fail; will deal with this later. */
+    fs->conn = new aws_bucket_conn(aws_config);
+    fs->file_system.terminate = s3_fs_terminate;
+
+    /* TODO: Move these into tests. Just testing here temporarily to show all functions work. */
+    {
+        /* List S3 buckets. */
+        std::vector<std::string> buckets;
+        if (fs->conn->list_buckets(buckets)) {
+            std::cout << "All buckets under my account:" << std::endl;
+            for (const std::string &bucket : buckets) {
+                std::cout << "  * " << bucket << std::endl;
+            }
+            std::cout << std::endl;
+        }
+
+        /* Have at least one bucket to use. */
+        if (buckets.size() >= 1) {
```

Comment:
You can use the function `empty()` instead of the `size()`, see the docs for [reference](https://www.cplusplus.com/reference/vector/vector/empty/).